signac/synced_collections/numpy_utils.py

```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Define utilities for handling NumPy arrays."""
+
+
+class NumpyConversionWarning(UserWarning):
+    """Warning that NumPy data was converted."""
+
+
+try:
+    import warnings
+
+    import numpy
+
+    NUMPY_CONVERSION_WARNING = (
+        "Any numpy types provided will be transparently converted to the "
+        "closest base Python equivalents."
+    )
+
+    def _convert_numpy(data):
+        """Convert a numpy data types to the corresponding base data types.
+
+        0d numpy arrays and numpy scalars are converted to their corresponding
+        primitive types, while other numpy arrays are converted to lists. If data
+        not a numpy data type, this function is a no-op.
+        """
+        if isinstance(data, numpy.ndarray):
+            # tolist will return a scalar for 0d arrays, so there's no need to
+            # special-case that check. 1-element 1d arrays should remain
+            # arrays, i.e. np.array([1])->[1], not 1.
+            warnings.warn(NUMPY_CONVERSION_WARNING, NumpyConversionWarning)
+            return data.tolist()
+        elif isinstance(data, (numpy.number, numpy.bool_)):
+            warnings.warn(NUMPY_CONVERSION_WARNING, NumpyConversionWarning)
+            return data.item()
+        return data
+
+    def _is_nonscalar_numpy_array(data):
+        """Check if an object is a nonscalar numpy array.
+
+        0d numpy arrays are not considered numpy arrays by this function. This
+        behavior is in line with typical requirements in the synced collections
+        framework, since >0d arrays are mapped to (synced) lists while 0d
+        arrays are mapped to scalars.
+
+        Returns
+        -------
+        bool
+            Whether or not the input is a numpy array.
+        """
+        return isinstance(data, numpy.ndarray) and data.shape != ()
+
+    def _is_numpy_scalar(data, allow_zero_d_array=False):
+        """Check if an object is a numpy scalar.
+
+        In certain cases a 0d numpy array is an acceptable surrogate for a scalar,
```

Comment:
This is a very good question, and actually leads to a more complicated issue. The immediate answer to your question is that, everywhere that you expect 0d arrays to work as scalars, they do, because we first call `_convert_numpy` in all of those cases. So actually this function is never called on a 0d array, and that entire branch of this function can be removed.

The thornier issue is _why_ this works. Right now, the reason is that `_convert_numpy` is being called by the `AbstractTypeResolver` internally, which means that 0d arrays are already being converted to scalars before any of the places where this function would be called. I use resolvers code base without making this assumption, so the `_is_numpy_scalar` function is being called in places where I anticipated having to handle 0d arrays, but then found that that case never arose.

The problem is that the resolver is basically a dictionary mapping from types->strings. While it uses callables to establish the mapping the first time, afterwards it does a simple dict lookup based on the type. This means that if the first numpy array it sees is >0d, it will forever map 0d numpy arrays to the same thing as other numpy arrays (and vice versa). Currently, this works because the resolver internally converts numpy arrays, so the resulting types end up being what we want them to be.

The easiest solution for now is to simply document this behavior of the class with respect to numpy arrays. I only think it's really an issue if someone tried to use it out of the context of synced collections, in which case this behavior might be unexpected. Frankly I don't _like_ that it handles this internally, but I don't see a good alternative. If you can think of a better way to do this then I'm all for it, otherwise I'd vote to punt in the interest of getting this merged. I'd like a cleaner solution, but it's a more subtle issue that requires some thinking and I'm bandwidth-limited so I'm not going to spend the time unless an easy answer comes to me.

Reply:
> So actually this function is never called on a 0d array, and that entire branch of this function can be removed.

...so can we remove it or not? I read the last 3 paragraphs 3 times but I never quite understood what you're getting at wrt how we should act here. Ultimately I'm fine with just dropping this conversation and letting it be, if that's easiest and lets us make forward progress. Resolve this however you see fit (whether with some change or no action).

Reply:
Sorry, I rambled. I've resolved this in a way that makes me slightly happier. Yes, I removed this branch entirely; this function will no longer return true for 0d numpy arrays. I also added an optional preprocessor argument to the AbstractTypeResolver, allowing places that need it to specify that the input data needs to be converted in some way prior to type resolution, avoiding any implicit behavior wrt numpy.